CachedPoints duplication and related fixes

skulonen opened the pull request:
# Description

**What?**

`CachedPoints` was duplicated in the delayed feedback update because that was the easiest way to store revealed and unrevealed exercise results simultaneously. This PR fixes the duplication.

Additionally, the PR contains several related bug fixes:
- `InspectSubmitterView` and `NextUnassessedSubmitterView` used the student version of the cache, which means they would redirect the user to the submitter's last submission if delayed feedback was enabled, even though staff members are supposed to be able to know which submission is the best one.
- The "my submissions" list would evaluate the exercise's reveal rule for each submission separately. This was unnecessary because the result would be the same for each submission. This has been optimized.
- The cache key `best_submission` would reveal the best submission through the student cache before feedback was supposed to be revealed. It is now reverted to the last submission in the student cache when feedback is not revealed.
- Some comments were outdated after the "delayed feedback" and "force exercise points" updates.
- The cache keys `feedback_revealed` and `feedback_reveal_time` were missing from submissions.
- The helper functions `has_more_points` and `is_newer` now accept a submission as the second argument, instead of the current cache entry. Because of this, `submission_date` (exercise key) no longer has to be stored in cache.

**Why?**

The duplicated cache took some extra memory. In addition, it would have to be generated twice since there were two versions of it, so this is also a performance improvement.

**How?**

`CachedPoints` now generates the staff and student cache immediately after one another, using the same query results for both passes. Because of this, the generation time is reduced. After both versions have been generated, they are merged by recursively comparing the values in the caches. When there is a conflicting value (e.g. points are revealed in staff cache but not in student cache), a tuple with both values is stored. This completely eliminates the duplication if a course doesn't use delayed feedback at all, and greatly reduces it when it does since only the differing keys have duplicated values. In the `CachedPoints` initializer, only one set of these duplicate keys is left in the dict, while the other set is removed. This has a small runtime cost (~4-5 ms), but makes it easier to use the cache in views and templates (the caller doesn't have to check which value should be used every time).

The "my submissions" list was optimized by allowing an `is_revealed` argument to be passed to the `points_badge` template tag. If it is provided, the reveal rule will not be evaluated by the template tag.

Fixes #887


# Testing

**Remember to add or update unit tests for new features and changes.**

* How to [test your changes in A-plus](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations)
* How to [test accessibility](https://wiki.aalto.fi/display/EDIT/How+to+check+the+accessibility+of+pull+requests)


**What type of test did you run?**

- [ ] Accessibility test using the [WAVE](https://wave.webaim.org/extension/) extension.
- [x] Django unit tests.
- [x] Selenium tests.
- [ ] Other test. *(Add a description below)*
- [x] Manual testing.

Tested manually that the cache still returns the expected values. In addition, the amount of memory taken by the cache was measured by accessing memcached via telnet.

**Did you test the changes in**

- [ ] Chrome
- [x] Firefox
- [ ] This pull request cannot be tested in the browser.

**Think of what is affected by these changes and could become broken**

# Translation

- [ ] Did you modify or add new strings in the user interface? ([Read about how to create translation](https://github.com/apluslms/a-plus/tree/master/doc#running-tests-and-updating-translations))

# Programming style

- [x] Did you follow our [style guides](https://apluslms.github.io/contribute/styleguides/)?

# Have you updated the README or other relevant documentation?

- [ ] documents inside the doc directory.
- [ ] README.md.
- [ ] Aplus Manual.
- [ ] Other documentation (mention below which documentation).

# Is it Done?

- [ ] Reviewer has finished the code review
- [ ] After the review, the developer has made changes accordingly
- [ ] Customer/Teacher has accepted the implementation of the feature

*Clean up your git commit history before submitting the pull request!*

skulonen commented:
This PR supersedes #907.

skulonen commented:
The current implementation stores duplicated values as tuples, and duplicated values are identified based on their type `tuple` in the `_extract_tuples` method. This works because there are no other tuples stored in the cache. But should we still use some other way to identify duplicated values, so we could theoretically store other tuples in the cache? Duplicated keys like in the original PR? A custom class that derives from `tuple`?

Mankro reviewed (changes requested):
Fantastic work! I wonder if we could optimize the memory consumption. I'm not sure if the current way of turning large QuerySets into lists is the most efficient solution.

In addition, the new dictionaries `final_submissions` and `last_submissions` consume some memory. You added them because the submissions are not grouped by the exercise and thus, the submissions loop may jump between exercises back and forth. What if the submissions database query grouped the submissions by the exercise so that you could loop over all submissions in one exercise at once? You wouldn't need to remember all best submissions in the whole course. What do you think? (The last submission would also be obvious if the submissions were sorted, though sorting has a significant cost too.)

Mankro commented:
> The current implementation stores duplicated values as tuples, and duplicated values are identified based on their type `tuple` in the `_extract_tuples` method. This works because there are no other tuples stored in the cache. But should we still use some other way to identify duplicated values, so we could theoretically store other tuples in the cache? Duplicated keys like in the original PR? A custom class that derives from `tuple`?

I think using tuples is fine. There is currently no obvious need to store tuples as "normal data" in the cache, hence using tuples for a special purpose is perfectly fine. Any additional extra identification field would just consume more memory.

skulonen commented on exercise/cache/points.py:
```diff
@@ -167,62 +228,78 @@ def r_augment(children: List[Dict[str, Any]]) -> None:
                     'unofficial': unofficial,
                     'date': submission.submission_time,
                     'url': submission.get_url('submission-plain'),
+                    'feedback_revealed': True,
+                    'feedback_reveal_time': None,
```
Good point and yes, they are the same between all submissions. The main (and I think only) reason I decided to store these for every submission back when I implemented delayed feedback, was `points_badge`. `points_badge` is designed to work with three alternative input types (see `_points_data`): a `UserExerciseSummary` object, a `Submission` object or a CachedPoints dict. The last option does not differentiate between different types of cache entries - it may be an exercise, a submission, a module or a category. If we got rid of these values in the submission dict, the submission dropdowns on the course front page would not know if feedback is revealed or not. We would need extra logic that passes these values to `points_badge` along with the submission dict itself.

Note that these aren't the only values that are redundantly stored for each submission. `max_points` is also the same for all submissions. And this:
```python
'submission_count': 1, # to fool points badge
```

Mankro replied:
That makes sense. Though, you have already added the `is_revealed` parameter to the `points_badge` template tag, but the parameter isn't always used (like in the course's Your points page). Let's not change this right now. We can leave this as is. It would be possible to avoid repeating these fields for every submission, but the changes would take too much time.

skulonen commented on exercise/cache/points.py:
```diff
@@ -73,6 +76,68 @@ def _generate_data(
             data: Optional[Dict[str, Any]] = None,
             ) -> Dict[str, Any]:
+        # Perform all database queries before generating the cache.
+        exercises = list(
```
The memory consumption is a good point. I didn't consider that, and instead only optimized for minimal execution time and minimal number of database queries. The generation code seemed to work ever so slightly faster when using `list()`, so I opted to use it. In addition, I logged all points when the generation code executes a query, and it seemed that when not using `list()`, some queries were executed during the second pass (when `is_staff=False`) because of reveal rules. I thought it might be more efficient to do all database queries upfront.

>I was under the impression that QuerySets cache the results inside the QuerySet object and you could iterate over the same QuerySet twice while loading the data only once from the database. I am not sure, though.

>Then again, is the QuerySet's cache actually in memory, thus either the memory is consumed by this list variable or a QuerySet variable?

Yes, the results are cached in the `QuerySet` object, so I would imagine either the list or the queryset would consume the memory. I'm not sure if it's the same amount of memory, though, since I didn't measure. It might be that the queryset's cache is somehow more optimized?

Mankro replied:
According to your test, the QuerySet object consumes even a little more memory than the simple list. Let's keep the list as is.

skulonen commented on exercise/cache/points.py:
```diff
@@ -73,6 +76,68 @@ def _generate_data(
+                .filter(exercise__course_module__course_instance=instance)
+                .select_related()
+                .prefetch_related('exercise__parent', 'notifications')
+                .only('id', 'exercise', 'submission_time', 'status', 'grade', 'force_exercise_points')
```
Reverse relationship names can't be used in `only()`. I tested and the code queries notifications only once when generating the cache.

skulonen commented on exercise/cache/points.py:
```diff
@@ -415,6 +491,44 @@ def submission_ids(
+
+    def _extract_tuples(self, value, tuple_index, parent_container=None, parent_key=None):
+        """
+        Finds tuples within a data structure, end replaces them with the value
```
Fixed.

skulonen commented:
>In addition, the new dictionaries final_submissions and last_submissions consume some memory.

Probably not very much. The submission objects are already stored in the `submissions` list. Since dicts store things by reference, these dicts will only map some integers to some memory addresses.

>You added them because the submissions are not grouped by the exercise and thus, the submissions loop may jump between exercises back and forth.

I added `final_submissions` so that I could pass the current final submission to the `is_better_than` functions, since they now take a submission as the second argument. Previously, the exercise cache entry was passed to those functions, which is why `submission_date` needed to be stored in the exercise cache entry. I added `last_submissions` so that I could set `best_submission` to point to the last submission in case the reveal rule evaluates to false.

>What if the submissions database query grouped the submissions by the exercise so that you could loop over all submissions in one exercise at once? You wouldn't need to remember all best submissions in the whole course. What do you think?

Sure. In that case we wouldn't need to store the final submissions as a dict, and instead it would only be one pointer to one submission object. `last_submissions` would still be needed because reveal rules are evaluated after submissions have been looped. Unless we change the logic so that a single exercise's reveal rule is handled right after after all of its submissions have been looped.

>though sorting has a significant cost too

The submissions are already sorted because of the `ordering` field in the Submission Meta class:

```sql
ORDER BY "exercise_submission"."id" DESC
```

skulonen commented:
I have now implemented the previously discussed change to the submissions loop. Commit: https://github.com/apluslms/a-plus/pull/911/commits/579fa02cf1115aa97b041458fb5989b2173ae28b

The main changes are:
- The submissions list is now ordered by exercise and then descending submission time.
- No more `final_submissions` and `last_submissions` dicts. Now we have `final_submission` and `last_submission` variables instead.
- Some variables are only updated when a new exercise is encountered (`exercise`, `entry`, `is_better_than`).
- Reveal rules are now checked when the last submission of the exercise has been handled (see `check_reveal_rule` function). Previously, they were handled in their own loop. This has the side effect that reveal rules are no longer checked for exercises that have no submissions, so their score will be shown as "0" instead of "?". This can be seen as a good or a bad thing.
- Deviations are now looped before submissions. This is because we have to know the deviations before reveal rules can be evaluated.

To be honest, I'm not sure if this is better or worse than the original version. In a way, I think the original version seemed more structured because each part was handled in its own, neat block (submissions, then deviations, then reveal rules). Bet now they're a bit intertwined. On the other hand, this seems to be a bit more efficient because a) we don't need to store the two dicts, b) `entry` does not have to resolved for each exercise. The difference is very small, though.

Feel free to tell me what you think. We can keep or revert this commit.

Mankro commented on exercise/cache/points.py:
```diff
@@ -22,28 +29,38 @@
-        return is_newer(submission, current_entry)
-    return submission.grade >= current_entry['points']
+# Type for the submission compare functions.
+IsBetterThan = Callable[[Submission, Optional[Submission]], bool]
```
Sometimes I have seen `if TYPE_CHECKING:` before listing static types. I'm just curious if this line would be a typical place to use the `if` or not.

(I am still not very familiar with static typing.)

skulonen replied:
Could be done either way. The reason `if TYPE_CHECKING` is usually used is when we want to do things that have no runtime effect. For example:
- For type hinting purposes, importing classes that only exist in the `django-types` type library but not in Django. (`RelatedManager`)
- For type hinting purposes, importing classes that would cause a cyclical import error at runtime.
- Type hinting model fields that Django adds at runtime (id, reverse relationships).

There isn't really a reason to use it here, but it could still be done because it kind of makes sense - `IsBetterThan` has no runtime purpose, it only exists so I could add a type to the `is_better_than` variable, which was initialized to `None`.

skulonen replied:
I actually decided to remove this `IsBetterThan` type and other initial types from the submission loop variables because they can be inferred by the IDE, and it was technically wrong to initialize them to `None` while annotating them with another type: https://github.com/apluslms/a-plus/pull/911/commits/723b6f8d18be7dea07f4599cb96a7a68ccfde27a.

Mankro replied:
Sounds good!

Mankro commented on exercise/cache/points.py:
```diff
@@ -133,21 +210,103 @@ def r_augment(children: List[Dict[str, Any]]) -> None:
+                        self.dirty = True
+                        continue
+                    entry = tree[-1]
+                    if submission.exercise.grading_mode == BaseExercise.GRADING_MODE.BEST:
```
`submission.exercise` could be just `exercise` here.

skulonen replied:
Fixed

Mankro commented on exercise/cache/points.py:
```diff
@@ -133,21 +210,103 @@ def r_augment(children: List[Dict[str, Any]]) -> None:
+                    'feedback_reveal_time': reveal_time,
+                })
+
+                for submission in entry['submissions']:
```
Looking at this loop, I started to think that this data could have been inserted correctly originally (the line 314 `entry['submissions'].append({`). The reveal rule can be defined with the exercise and the exercise is known in the beginning of the submissions loop on line 286. The reveal rule could be used to hide the points when submissions are inserted to the `entry['submissions']` list.

skulonen replied:
No, the reveal rule must be evaluated after all submissions have been iterated. This is because the `COMPLETION` reveal trigger depends on the submission count and the final points of the exercise. If that trigger option didn't exist, we could do what you said and evaluate the reveal rule right when a new exercise is encountered.

Mankro replied:
Right you are! Then you may fix the commit history and merge we shall!

Mankro reviewed (approved):
I like the new approach. :+1:  Indeed, this probably affects performance less than what I maybe initially thought, but it's done now.

You said you didn't have time to test this much, so could you still spend a moment testing this? I read the code and it feels correct to me.

I commented on something below that could a very easy, small improvement. If you agree, you could still change that.

You commented on exercises with no submissions, using delayed feedback, and the student sees 0 points instead of ? points:
it makes perfect sense that with no submissions, you get zero points. It's good to show that in the UI. So no problem with that!

skulonen commented:
I realized that `_generate_data_internal` no longer needs to receive the `exercises` list as a parameter, so I removed that.

I have now cleaned the commit history.

skulonen commented:
By the way, I think CachedContent could benefit from some optimization too. Generating it takes about 200 ms in the A+ manual course, and according to Django Debug Toolbar, it repeats a lot of queries unnecessarily. On the other hand, CachedContent is invalidated much more rarely than CachedPoints.

Mankro reviewed (approved):
Excellent work! Now I merge.

Mankro commented:
> By the way, I think CachedContent could benefit from some optimization too. Generating it takes about 200 ms in the A+ manual course, and according to Django Debug Toolbar, it repeats a lot of queries unnecessarily. On the other hand, CachedContent is invalidated much more rarely than CachedPoints.

New issue #912
